from typing import ClassVar
from typing_extensions import TypeAlias, Literal
from datetime import date
from uniprot_rest.types.util import UniprotMixin
from dataclasses import dataclass
Identity: TypeAlias = Literal['1.0', '0.9', '0.5']

@dataclass
class Uniref(UniprotMixin):
    id: str
    name: str
    identity: Identity
    count: int
    length: int
    created: date
    uniprot_id: str
    upi: str
    taxonomy_name: str
    cluster: str

@dataclass
class Uniparc(UniprotMixin):
    upi: str
    uniprotkb: str
    isoform: str
    upid: str
    taxonomy_name: str
    gene: str
    protein: str
    database: str
    active: str
    checksum: str
    length: int
    dbid: str
    feature_id: str
    proteomecomponent: str
Rank: TypeAlias = Literal['SUPERKINGDOM', 'KINGDOM', 'SUBKINGDOM', 'SUPERPHYLUM', 'PHYLUM', 'SUBPHYLUM', 'SUPERCLASS', 'CLASS', 'SUBCLASS', 'INFRACLASS', 'COHORT', 'SUBCOHORT', 'SUPERORDER', 'ORDER', 'SUBORDER', 'INFRAORDER', 'PARVORDER', 'SUPERFAMILY', 'FAMILY', 'SUBFAMILY', 'TRIBE', 'SUBTRIBE', 'GENUS', 'SUBGENUS', 'SPECIES_GROUP', 'SPECIES_SUBGROUP', 'SPECIES', 'SUBSPECIES', 'VARIETAS', 'FORMA', 'NO_RANK']

@dataclass
class Taxonomy(UniprotMixin):
    tax_id: int
    scientific: str
    common: str
    mnemonic: str
    rank: Rank
    strain: str
    host: int
    linked: bool
    parent: str
    ancestor: str
Existence: TypeAlias = Literal['1', '2', '3', '4', '5']
Organelle: TypeAlias = Literal['mitochondrion', 'plastid', 'chloroplast', 'cyanelle', 'apicoplast', 'organellar chromatophore', 'non-photosynthetic plastid', 'nucleomorph', 'hydrogenosome']

@dataclass
class Uniprotkb(UniprotMixin):
    accession_field: str
    id_field: str
    protein_name_field: str
    gene_field: str
    organism_name_field: str
    taxonomy_name: str
    virus_host_name: str
    existence: Existence
    ec: str
    cc_cofactor_chebi: str
    ccev_cofactor_chebi: str
    cc_cofactor_note: str
    ccev_cofactor_note: str
    cc_bpcp: str
    ccev_bpcp: str
    cc_bpcp_absorption: str
    ccev_bpcp_absorption: str
    cc_bpcp_kinetics: str
    ccev_bpcp_kinetics: str
    cc_bpcp_ph_dependence: str
    ccev_bpcp_ph_dependence: str
    cc_bpcp_redox_potential: str
    ccev_bpcp_redox_potential: str
    cc_bpcp_temp_dependence: str
    ccev_bpcp_temp_dependence: str
    cc_catalytic_activity_field: str
    ccev_catalytic_activity: str
    cc_activity_regulation: str
    ccev_activity_regulation: str
    cc_function: str
    ccev_function: str
    cc_caution: str
    ccev_caution: str
    ft_sites: str
    ftlen_sites: int
    ftev_sites: str
    ft_act_site: str
    ftlen_act_site: int
    ftev_act_site: str
    ft_metal: str
    ftlen_metal: int
    ftev_metal: str
    ft_binding: str
    ftlen_binding: int
    ftev_binding: str
    ft_site: str
    ftlen_site: int
    ftev_site: str
    ft_ca_bind: str
    ftlen_ca_bind: int
    ftev_ca_bind: str
    ft_dna_bind: str
    ftlen_dna_bind: int
    ftev_dna_bind: str
    ft_np_bind: str
    ftlen_np_bind: int
    ftev_np_bind: str
    cc_pathway: str
    ccev_pathway: str
    cc_miscellaneous: str
    ccev_miscellaneous: str
    cc_scl_term_field: str
    ccev_scl_term: str
    cc_scl_note: str
    ccev_scl_note: str
    ft_transmem: str
    ftlen_transmem: int
    ftev_transmem: str
    ft_topo_dom: str
    ftlen_topo_dom: int
    ftev_topo_dom: str
    ft_intramem: str
    ftlen_intramem: int
    ftev_intramem: str
    cc_disease: str
    ccev_disease: str
    cc_allergen: str
    ccev_allergen: str
    cc_toxic_dose: str
    ccev_toxic_dose: str
    cc_biotechnology: str
    ccev_biotechnology: str
    cc_pharmaceutical: str
    ccev_pharmaceutical: str
    cc_disruption_phenotype: str
    ccev_disruption_phenotype: str
    ft_mutagen: str
    ftlen_mutagen: int
    ftev_mutagen: str
    cc_ptm: str
    ccev_ptm: str
    ft_mod_res: str
    ftlen_mod_res: int
    ftev_mod_res: str
    ft_lipid: str
    ftlen_lipid: int
    ftev_lipid: str
    ft_carbohyd: str
    ftlen_carbohyd: int
    ftev_carbohyd: str
    ft_disulfid: str
    ftlen_disulfid: int
    ftev_disulfid: str
    ft_crosslnk: str
    ftlen_crosslnk: int
    ftev_crosslnk: str
    ft_molecule_processing: str
    ftlen_molecule_processing: int
    ftev_molecule_processing: str
    ft_chain: str
    ftlen_chain: int
    ftev_chain: str
    ft_init_met: str
    ftlen_init_met: int
    ftev_init_met: str
    ft_peptide: str
    ftlen_peptide: int
    ftev_peptide: str
    ft_signal: str
    ftlen_signal: int
    ftev_signal: str
    ft_propep: str
    ftlen_propep: int
    ftev_propep: str
    ft_transit: str
    ftlen_transit: int
    ftev_transit: str
    cc_developmental_stage: str
    ccev_developmental_stage: str
    cc_induction: str
    ccev_induction: str
    cc_tissue_specificity: str
    ccev_tissue_specificity: str
    interactor: str
    cc_subunit: str
    ccev_subunit: str
    structure_threed: bool
    ft_secstruct: str
    ftlen_secstruct: int
    ftev_secstruct: str
    ft_helix: str
    ftlen_helix: int
    ftev_helix: str
    ft_turn: str
    ftlen_turn: int
    ftev_turn: str
    ft_strand: str
    ftlen_strand: int
    ftev_strand: str
    mass_range: int
    length_range: int
    cc_ap: str
    ccev_ap: str
    cc_ap_apu: str
    ccev_ap_apu: str
    cc_ap_as: str
    ccev_ap_as: str
    cc_ap_ai: str
    ccev_ap_ai: str
    cc_ap_rf: str
    ccev_ap_rf: str
    cc_sequence_caution: str
    ccev_sequence_caution: str
    cc_sc_framesh: str
    cc_sc_einit: str
    cc_sc_eterm: str
    cc_sc_epred: str
    cc_sc_etran: str
    cc_sc_misc: str
    ccev_sc_misc: str
    cc_mass_spectrometry: str
    ccev_mass_spectrometry: str
    cc_polymorphism: str
    ccev_polymorphism: str
    cc_rna_editing: str
    ccev_rna_editing: str
    ft_variants: str
    ftlen_variants: int
    ftev_variants: str
    ft_variant: str
    ftlen_variant: int
    ftev_variant: str
    ft_var_seq: str
    ftlen_var_seq: int
    ftev_var_seq: str
    ft_non_std: str
    ftlen_non_std: int
    ftev_non_std: str
    ft_non_ter: str
    ftlen_non_ter: int
    ftev_non_ter: str
    ft_non_cons: str
    ftlen_non_cons: int
    ftev_non_cons: str
    ft_conflict: str
    ftlen_conflict: int
    ftev_conflict: str
    ft_unsure: str
    ftlen_unsure: int
    ftev_unsure: str
    ft_positional: str
    ftlen_positional: int
    ftev_positional: str
    fragment: bool
    organelle: Organelle
    precursor: bool
    tissue: str
    strain: str
    plasmid: str
    transposon: str
    ft_domain: str
    ftlen_domain: int
    ftev_domain: str
    cc_domain: str
    ccev_domain: str
    family: str
    ft_coiled: str
    ftlen_coiled: int
    ftev_coiled: str
    ft_compbias: str
    ftlen_compbias: int
    ftev_compbias: str
    ft_motif: str
    ftlen_motif: int
    ftev_motif: str
    ft_region: str
    ftlen_region: int
    ftev_region: str
    ft_repeat: str
    ftlen_repeat: int
    ftev_repeat: str
    cc_similarity: str
    ccev_similarity: str
    ft_zn_fing: str
    ftlen_zn_fing: int
    ftev_zn_fing: str
    xref_any: str
    xref_embl: str
    xref_ccds: str
    xref_pir: str
    xref_refseq: str
    xref_pdb: str
    xref_pdbsum: str
    xref_pcddb: str
    xref_sasbdb: str
    xref_bmrb: str
    xref_smr: str
    xref_alphafolddb: str
    xref_biogrid: str
    xref_complexportal: str
    xref_corum: str
    xref_dip: str
    xref_elm: str
    xref_intact: str
    xref_mint: str
    xref_string: str
    xref_bindingdb: str
    xref_chembl: str
    xref_drugbank: str
    xref_guidetopharmacology: str
    xref_swisslipids: str
    xref_drugcentral: str
    xref_allergome: str
    xref_cazy: str
    xref_esther: str
    xref_imgt_gene_db: str
    xref_merops: str
    xref_moondb: str
    xref_moonprot: str
    xref_clae: str
    xref_peroxibase: str
    xref_rebase: str
    xref_tcdb: str
    xref_unilectin: str
    xref_carbonyldb: str
    xref_depod: str
    xref_glyconnect: str
    xref_glygen: str
    xref_iptmnet: str
    xref_phosphositeplus: str
    xref_swisspalm: str
    xref_unicarbkb: str
    xref_metosite: str
    xref_biomuta: str
    xref_dmdm: str
    xref_dbsnp: str
    xref_compluyeast_twodpage: str
    xref_dosac_cobs_twodpage: str
    xref_ogp: str
    xref_reproduction_twodpage: str
    xref_swiss_twodpage: str
    xref_ucd_twodpage: str
    xref_world_twodpage: str
    xref_cptac: str
    xref_epd: str
    xref_maxqb: str
    xref_paxdb: str
    xref_peptideatlas: str
    xref_pride: str
    xref_promex: str
    xref_proteomicsdb: str
    xref_topdownproteomics: str
    xref_jpost: str
    xref_massive: str
    xref_dnasu: str
    xref_abcd: str
    xref_antibodypedia: str
    xref_cptc: str
    xref_ensembl: str
    xref_ensemblbacteria: str
    xref_ensemblfungi: str
    xref_ensemblmetazoa: str
    xref_ensemblplants: str
    xref_ensemblprotists: str
    xref_geneid: str
    xref_gramene: str
    xref_kegg: str
    xref_mane_select: str
    xref_patric: str
    xref_ucsc: str
    xref_vectorbase: str
    xref_wbparasite: str
    xref_wbparasitetranscriptprotein: str
    xref_arachnoserver: str
    xref_araport: str
    xref_cgd: str
    xref_conoserver: str
    xref_ctd: str
    xref_dictybase: str
    xref_disgenet: str
    xref_echobase: str
    xref_euhcvdb: str
    xref_veupathdb: str
    xref_flybase: str
    xref_genecards: str
    xref_genereviews: str
    xref_hgnc: str
    xref_hpa: str
    xref_legiolist: str
    xref_leproma: str
    xref_maizegdb: str
    xref_malacards: str
    xref_mgi: str
    xref_mim: str
    xref_niagads: str
    xref_nextprot: str
    xref_opentargets: str
    xref_orphanet: str
    xref_pharmgkb: str
    xref_pombase: str
    xref_pseudocap: str
    xref_rgd: str
    xref_sgd: str
    xref_tair: str
    xref_tuberculist: str
    xref_vgnc: str
    xref_wormbase: str
    xref_xenbase: str
    xref_zfin: str
    xref_eggnog: str
    xref_genetree: str
    xref_hogenom: str
    xref_inparanoid: str
    xref_ko: str
    xref_oma: str
    xref_orthodb: str
    xref_phylomedb: str
    xref_treefam: str
    xref_biocyc: str
    xref_brenda: str
    xref_reactome: str
    xref_sabio_rk: str
    xref_signalink: str
    xref_signor: str
    xref_unipathway: str
    xref_plantreactome: str
    xref_pathwaycommons: str
    xref_chitars: str
    xref_evolutionarytrace: str
    xref_genewiki: str
    xref_genomernai: str
    xref_phi_base: str
    xref_pro: str
    xref_pharos: str
    xref_rnact: str
    xref_biogrid_orcs: str
    xref_bgee: str
    xref_cleanex: str
    xref_collectf: str
    xref_expressionatlas: str
    xref_genevisible: str
    xref_cdd: str
    xref_genethreed: str
    xref_hamap: str
    xref_ideal: str
    xref_interpro: str
    xref_panther: str
    xref_pfam: str
    xref_pirsf: str
    xref_prints: str
    xref_prodom: str
    xref_sfld: str
    xref_smart: str
    xref_supfam: str
    xref_tigrfams: str
    xref_prosite: str
    xref_disprot: str
    xref_go: str
    xref_proteomes: str
    database: str
    cc_webresource: str
    ccev_webresource: str
    date_created: date
    date_modified: date
    date_sequence_modified: date
    go_field: str
    go_evidence: str
    chebi_field: str
    inchikey_field: str
    keyword_field: str
    lit_author: str
    lit_journal: str
    lit_pubdate: date
    lit_pubmed: str
    lit_title: str
    lit_citation_id: str
    computational_pubmed_id: str
    community_pubmed_id: str
    proteome: str
    proteome_component: str
    cited_for: str
    reviewed: bool
    active: bool
    uniref_cluster_50: str
    uniref_cluster_90: str
    uniref_cluster_100: str
    uniparc: str
    field_map: ClassVar[dict] = {'structure_threed': 'structure_3d', 'xref_imgt_gene_db': 'xref_imgt_gene-db', 'xref_compluyeast_twodpage': 'xref_compluyeast-2dpage', 'xref_dosac_cobs_twodpage': 'xref_dosac-cobs-2dpage', 'xref_reproduction_twodpage': 'xref_reproduction-2dpage', 'xref_swiss_twodpage': 'xref_swiss-2dpage', 'xref_ucd_twodpage': 'xref_ucd-2dpage', 'xref_world_twodpage': 'xref_world-2dpage', 'xref_mane_select': 'xref_mane-select', 'xref_sabio_rk': 'xref_sabio-rk', 'xref_phi_base': 'xref_phi-base', 'xref_biogrid_orcs': 'xref_biogrid-orcs', 'xref_genethreed': 'xref_gene3d'}
Proteome_type: TypeAlias = Literal['1', '2', '3', '4']
Cpd: TypeAlias = Literal['1', '2', '3', '4', '5', '6']

@dataclass
class Proteomes(UniprotMixin):
    upid: str
    proteome_type: Proteome_type
    organism_name: str
    taxonomy_name: str
    genome_accession: str
    genome_assembly: str
    cpd: Cpd
    busco: int

@dataclass
class Locations(UniprotMixin):
    name: str
    id: str

@dataclass
class Unirule(UniprotMixin):
    protein_name_field: str
    gene_name_field: str
    organism_field: str
    taxonomy_field: str
    ec: str
    cc_cofactor: str
    cc_cofactor_note: str
    cc_catalytic_activity_field: str
    cc_activity_regulation: str
    cc_pathway: str
    cc_scl_term_field: str
    cc_scl_note: str
    cc_induction: str
    cc_domain: str
    family: str
    cc_similarity: str
    go_field: str
    keyword_field: str

@dataclass
class Diseases(UniprotMixin):
    name: str
    id: str

@dataclass
class Keywords(UniprotMixin):
    name: str
    keyword_id: str
    category: str

@dataclass
class Citations(UniprotMixin):
    title: str
    author: str
    journal: str
    published: str
    id: str
    doi: str

@dataclass
class Database(UniprotMixin):
    name: str
    id: str

@dataclass
class Arba(UniprotMixin):
    protein_name_field: str
    organism_field: str
    taxonomy_field: str
    ec: str
    cc_cofactor: str
    cc_catalytic_activity_field: str
    cc_activity_regulation: str
    cc_pathway: str
    cc_scl_term_field: str
    cc_scl_note: str
    cc_domain: str
    family: str
    cc_similarity: str
    keyword_field: str
__all__ = ['Uniref', 'Uniparc', 'Taxonomy', 'Uniprotkb', 'Proteomes', 'Locations', 'Unirule', 'Diseases', 'Keywords', 'Citations', 'Database', 'Arba']
