from typing import Union
from typing_extensions import TypeAlias, Literal
UnirefNamesTaxonomy: TypeAlias = Literal['id', 'name', 'common_taxon', 'common_taxonid', 'organism_id', 'organism']
UnirefSequences: TypeAlias = Literal['identity', 'length', 'sequence']
UnirefMiscellaneous: TypeAlias = Literal['types', 'members', 'count']
UnirefDateOf: TypeAlias = Literal['created',]
UnirefFields: TypeAlias = Union[UnirefNamesTaxonomy, UnirefSequences, UnirefMiscellaneous, UnirefDateOf]
UniparcNamesTaxonomy: TypeAlias = Literal['upi', 'gene', 'organism_id', 'organism', 'protein', 'proteome']
UniparcSequences: TypeAlias = Literal['checksum', 'length', 'sequence']
UniparcMiscellaneous: TypeAlias = Literal['accession',]
UniparcDateOf: TypeAlias = Literal['first_seen', 'last_seen']
UniparcFamilyDomains: TypeAlias = Literal['CDD', 'Gene3D', 'HAMAP', 'PANTHER', 'Pfam', 'PIRSF', 'PRINTS', 'PROSITE', 'SFLD', 'SMART', 'SUPFAM', 'TIGRFAMs']
UniparcFields: TypeAlias = Union[UniparcNamesTaxonomy, UniparcSequences, UniparcMiscellaneous, UniparcDateOf, UniparcFamilyDomains]
TaxonomyTaxonomy: TypeAlias = Literal['id', 'mnemonic', 'scientific_name', 'common_name', 'synonyms', 'other_names', 'rank', 'reviewed', 'lineage', 'strains', 'parent', 'hosts', 'links', 'statistics']
TaxonomyFields: TypeAlias = Union[TaxonomyTaxonomy,]
UniprotkbNamesTaxonomy: TypeAlias = Literal['accession', 'id', 'gene_names', 'gene_oln', 'gene_orf', 'gene_primary', 'gene_synonym', 'organism_name', 'organism_id', 'protein_name', 'xref_proteomes', 'lineage', 'lineage_ids', 'virus_hosts']
UniprotkbSequences: TypeAlias = Literal['cc_alternative_products', 'ft_var_seq', 'error_gmodel_pred', 'fragment', 'organelle', 'length', 'mass', 'cc_mass_spectrometry', 'ft_variant', 'ft_non_cons', 'ft_non_std', 'ft_non_ter', 'cc_polymorphism', 'cc_rna_editing', 'sequence', 'cc_sequence_caution', 'ft_conflict', 'ft_unsure', 'sequence_version']
UniprotkbFunction: TypeAlias = Literal['absorption', 'ft_act_site', 'ft_binding', 'ft_ca_bind', 'cc_catalytic_activity', 'cc_cofactor', 'ft_dna_bind', 'ec', 'cc_activity_regulation', 'cc_function', 'kinetics', 'ft_metal', 'ft_np_bind', 'cc_pathway', 'ph_dependence', 'redox_potential', 'rhea', 'ft_site', 'temp_dependence']
UniprotkbMiscellaneous: TypeAlias = Literal['annotation_score', 'cc_caution', 'keyword', 'keywordid', 'cc_miscellaneous', 'protein_existence', 'reviewed', 'tools', 'uniparc_id', 'comment_count', 'feature_count']
UniprotkbInteraction: TypeAlias = Literal['cc_interaction', 'cc_subunit']
UniprotkbExpression: TypeAlias = Literal['cc_developmental_stage', 'cc_induction', 'cc_tissue_specificity']
UniprotkbGeneOntologyGO: TypeAlias = Literal['go_p', 'go_c', 'go', 'go_f', 'go_id']
UniprotkbPathologyBiotech: TypeAlias = Literal['cc_allergen', 'cc_biotechnology', 'cc_disruption_phenotype', 'cc_disease', 'ft_mutagen', 'cc_pharmaceutical', 'cc_toxic_dose']
UniprotkbSubcellularLocation: TypeAlias = Literal['ft_intramem', 'cc_subcellular_location', 'ft_topo_dom', 'ft_transmem']
UniprotkbPTMProcessing: TypeAlias = Literal['ft_chain', 'ft_crosslnk', 'ft_disulfid', 'ft_carbohyd', 'ft_init_met', 'ft_lipid', 'ft_mod_res', 'ft_peptide', 'cc_ptm', 'ft_propep', 'ft_signal', 'ft_transit']
UniprotkbStructure: TypeAlias = Literal['structure_3d', 'ft_strand', 'ft_helix', 'ft_turn']
UniprotkbPublications: TypeAlias = Literal['lit_pubmed_id',]
UniprotkbDateOf: TypeAlias = Literal['date_created', 'date_modified', 'date_sequence_modified', 'version']
UniprotkbFamilyDomains: TypeAlias = Literal['ft_coiled', 'ft_compbias', 'cc_domain', 'ft_domain', 'ft_motif', 'protein_families', 'ft_region', 'ft_repeat', 'ft_zn_fing']
UniprotkbSequence: TypeAlias = Literal['xref_ccds', 'xref_embl', 'xref_pir', 'xref_refseq']
UniprotkbthreeDStructure: TypeAlias = Literal['xref_alphafolddb', 'xref_bmrb', 'xref_pcddb', 'xref_pdb', 'xref_pdbsum', 'xref_sasbdb', 'xref_smr']
UniprotkbProteinProteinInteraction: TypeAlias = Literal['xref_biogrid', 'xref_corum', 'xref_complexportal', 'xref_dip', 'xref_elm', 'xref_intact', 'xref_mint', 'xref_string']
UniprotkbChemistry: TypeAlias = Literal['xref_bindingdb', 'xref_chembl', 'xref_drugbank', 'xref_drugcentral', 'xref_guidetopharmacology', 'xref_swisslipids']
UniprotkbProteinFamilygroup: TypeAlias = Literal['xref_allergome', 'xref_cazy', 'xref_clae', 'xref_esther', 'xref_imgt_gene-db', 'xref_merops', 'xref_moondb', 'xref_moonprot', 'xref_peroxibase', 'xref_rebase', 'xref_tcdb', 'xref_unilectin']
UniprotkbPTM: TypeAlias = Literal['xref_carbonyldb', 'xref_depod', 'xref_glyconnect', 'xref_glygen', 'xref_metosite', 'xref_phosphositeplus', 'xref_swisspalm', 'xref_unicarbkb', 'xref_iptmnet']
UniprotkbGeneticVariation: TypeAlias = Literal['xref_biomuta', 'xref_dmdm', 'xref_dbsnp']
UniprotkbtwoDGel: TypeAlias = Literal['xref_compluyeast-2dpage', 'xref_dosac-cobs-2dpage', 'xref_ogp', 'xref_reproduction-2dpage', 'xref_swiss-2dpage', 'xref_ucd-2dpage', 'xref_world-2dpage']
UniprotkbProteomic: TypeAlias = Literal['xref_cptac', 'xref_epd', 'xref_massive', 'xref_maxqb', 'xref_pride', 'xref_paxdb', 'xref_peptideatlas', 'xref_promex', 'xref_proteomicsdb', 'xref_topdownproteomics', 'xref_jpost']
UniprotkbProtocolsAndMaterials: TypeAlias = Literal['xref_abcd', 'xref_antibodypedia', 'xref_cptc', 'xref_dnasu']
UniprotkbGenomeAnnotation: TypeAlias = Literal['xref_ensembl', 'xref_ensemblbacteria', 'xref_ensemblfungi', 'xref_ensemblmetazoa', 'xref_ensemblplants', 'xref_ensemblprotists', 'xref_geneid', 'xref_gramene', 'xref_kegg', 'xref_mane-select', 'xref_patric', 'xref_ucsc', 'xref_vectorbase', 'xref_wbparasite', 'xref_wbparasitetranscriptprotein']
UniprotkbOrganismSpecific: TypeAlias = Literal['xref_arachnoserver', 'xref_araport', 'xref_cgd', 'xref_ctd', 'xref_conoserver', 'xref_disgenet', 'xref_echobase', 'xref_flybase', 'xref_genecards', 'xref_genereviews', 'xref_hgnc', 'xref_hpa', 'xref_legiolist', 'xref_leproma', 'xref_mgi', 'xref_mim', 'xref_maizegdb', 'xref_malacards', 'xref_niagads', 'xref_opentargets', 'xref_orphanet', 'xref_pharmgkb', 'xref_pombase', 'xref_pseudocap', 'xref_rgd', 'xref_sgd', 'xref_tair', 'xref_tuberculist', 'xref_veupathdb', 'xref_vgnc', 'xref_wormbase', 'xref_xenbase', 'xref_zfin', 'xref_dictybase', 'xref_euhcvdb', 'xref_nextprot']
UniprotkbPhylogenomic: TypeAlias = Literal['xref_genetree', 'xref_hogenom', 'xref_inparanoid', 'xref_ko', 'xref_oma', 'xref_orthodb', 'xref_phylomedb', 'xref_treefam', 'xref_eggnog']
UniprotkbEnzymeAndPathway: TypeAlias = Literal['xref_brenda', 'xref_biocyc', 'xref_pathwaycommons', 'xref_plantreactome', 'xref_reactome', 'xref_sabio-rk', 'xref_signor', 'xref_signalink', 'xref_unipathway']
UniprotkbMiscellaneous: TypeAlias = Literal['xref_biogrid-orcs', 'xref_chitars', 'xref_evolutionarytrace', 'xref_genewiki', 'xref_genomernai', 'xref_phi-base', 'xref_pro', 'xref_pharos', 'xref_rnact']
UniprotkbGeneExpression: TypeAlias = Literal['xref_bgee', 'xref_cleanex', 'xref_collectf', 'xref_expressionatlas', 'xref_genevisible']
UniprotkbFamilyAndDomain: TypeAlias = Literal['xref_cdd', 'xref_disprot', 'xref_gene3d', 'xref_hamap', 'xref_ideal', 'xref_interpro', 'xref_panther', 'xref_pirsf', 'xref_prints', 'xref_prosite', 'xref_pfam', 'xref_prodom', 'xref_sfld', 'xref_smart', 'xref_supfam', 'xref_tigrfams']
UniprotkbFields: TypeAlias = Union[UniprotkbNamesTaxonomy, UniprotkbSequences, UniprotkbFunction, UniprotkbMiscellaneous, UniprotkbInteraction, UniprotkbExpression, UniprotkbGeneOntologyGO, UniprotkbPathologyBiotech, UniprotkbSubcellularLocation, UniprotkbPTMProcessing, UniprotkbStructure, UniprotkbPublications, UniprotkbDateOf, UniprotkbFamilyDomains, UniprotkbSequence, UniprotkbthreeDStructure, UniprotkbProteinProteinInteraction, UniprotkbChemistry, UniprotkbProteinFamilygroup, UniprotkbPTM, UniprotkbGeneticVariation, UniprotkbtwoDGel, UniprotkbProteomic, UniprotkbProtocolsAndMaterials, UniprotkbGenomeAnnotation, UniprotkbOrganismSpecific, UniprotkbPhylogenomic, UniprotkbEnzymeAndPathway, UniprotkbMiscellaneous, UniprotkbGeneExpression, UniprotkbFamilyAndDomain]
ProteomesNamesTaxonomy: TypeAlias = Literal['upid', 'organism', 'organism_id', 'components', 'mnemonic', 'lineage']
ProteomesMiscellaneous: TypeAlias = Literal['busco', 'cpd', 'genome_assembly', 'genome_representation', 'protein_count']
ProteomesFields: TypeAlias = Union[ProteomesNamesTaxonomy, ProteomesMiscellaneous]
LocationsSubcellularLocation: TypeAlias = Literal['id', 'name', 'definition', 'category', 'keyword', 'synonyms', 'content', 'gene_ontologies', 'note', 'references', 'links', 'is_a', 'part_of', 'statistics']
LocationsFields: TypeAlias = Union[LocationsSubcellularLocation,]
UniruleUniRule: TypeAlias = Literal['rule_id', 'template_entries', 'statistics', 'taxonomic_scope', 'annotation_covered', 'predicted_protein_name']
UniruleFields: TypeAlias = Union[UniruleUniRule,]
DiseasesDisease: TypeAlias = Literal['id', 'name', 'acronym', 'definition', 'alternative_names', 'cross_references', 'keywords', 'statistics']
DiseasesFields: TypeAlias = Union[DiseasesDisease,]
KeywordsKeyword: TypeAlias = Literal['id', 'name', 'definition', 'category', 'synonyms', 'gene_ontologies', 'links', 'children', 'parents', 'statistics']
KeywordsFields: TypeAlias = Union[KeywordsKeyword,]
CitationsLiterature: TypeAlias = Literal['id', 'doi', 'title', 'authors', 'authoring_group', 'journal', 'publication_date', 'first_page', 'last_page', 'volume', 'reference', 'lit_abstract', 'statistics']
CitationsFields: TypeAlias = Union[CitationsLiterature,]
DatabaseCrossReference: TypeAlias = Literal['id', 'name', 'abbrev', 'pubmed_id', 'doi_id', 'link_type', 'server', 'dbUrl', 'category', 'statistics']
DatabaseFields: TypeAlias = Union[DatabaseCrossReference,]
ArbaARBA: TypeAlias = Literal['rule_id', 'statistics', 'taxonomic_scope', 'annotation_covered']
ArbaFields: TypeAlias = Union[ArbaARBA,]
__all__ = ['UnirefFields', 'UniparcFields', 'TaxonomyFields', 'UniprotkbFields', 'ProteomesFields', 'LocationsFields', 'UniruleFields', 'DiseasesFields', 'KeywordsFields', 'CitationsFields', 'DatabaseFields', 'ArbaFields']
